### ISSUE TITLE
unix Domain socket creation failure

### ISSUE THREAD

**Issue by dragon512** [Microsoft/BashOnWindows#1083]
Been trying to do work with Apache trafficserver under this subsystem. ( First love the work guys.. 95% of what i need seem here only a few hickups one of then reported here). The issue at hand is that when I try to run some some top level programs in ATS such as traffic_manager I get an error such as:

[E. Mgmt] log ==> [TrafficManager] using root directory '/sb/basic-manager/ts'
FATAL: Can't open server lockfile '/sb/basic-manager/ts/var/trafficserver/server.lock' (Permission denied)

I should not see this error... 

To reproduce this in a simple way

1) git clone https://github.com/apache/trafficserver.git
2) sudo apt-get autoconf automake autotools-dev bison debhelper dh-apparmor flex gettext intltool-debian libbison-dev libcap-dev libexpat1-dev libfl-dev libpcre3-dev libpcrecpp0 libsigsegv2 libsqlite3-dev libssl-dev libtool m4 po-debconf tcl-dev tcl8.6-dev zlib1g-dev
3) cd traffic_server
4) autoreconf -if; ./configure --prefix /home/<user>/ats ; make all -j4; make install
5) cd ../ats/bin
6_ sudo ./traffic_manager

should see something like:
[E. Mgmt] log ==> [TrafficManager] using root directory '/home/jkenny/ats_master'
FATAL: Can't open server lockfile '/home/jkenny/ats/var/trafficserver/server.lock' (Permission denied)
[Sep 14 21:12:42.076] Manager {0x7f4584570780} NOTE: FATAL: Can't open server lockfile '/home/jkenny/ats/var/trafficserver/server.lock' (Permission denied)
[Sep 14 21:12:42.076] Manager {0x7f4584570780} NOTE: FATAL: Can't open server lockfile '/home/jkenny/ats/var/trafficserver/server.lock' (Permission denied)

This is a failure in creation of a unix domain socket.

Note on 6) normally we would not need to run sudo to run this command. however there is an call that fails on bash for windows that normally works on Ubuntu. you can see this via running ./traffic_server instead. The stack dump points to the Linux code in question. running sudo ./traffic_server works fine as a work around. This is a different issue BTW

However when needing to run to programs such as traffic_cop or traffic_manager we have failures in being able to start as traffic_manager uses UNIX domain sockets to communicate with traffic_server

**Comment by fpqc** [comment]
@dragon512 What Windows 10 build?  There were big improvements to sockets in Win10 insider build 14915.

**Comment by dragon512** [comment]
Sorry did not add that... build 14915.rs_prelease.160826-1902

**Comment by misenesi** [comment]
Hi @dragon512 ,

I tried to repro the issue following your steps (instead 3) I had to do cd trafficserver and install autoreconf as well for step 4), where I also tried to use '/home/ats' where there seems to be a typoo), but cannot get past step 4).

Could you please verify the repro steps?

Thank you!

**Comment by dragon512** [comment]
sorry for the typos .... some of 4) was messed up as some text did not copy over :-(

let me tweak it a little..

1) git clone https://github.com/apache/trafficserver.git
2) sudo apt-get autoconf automake autotools-dev bison debhelper dh-apparmor flex gettext intltool-debian libbison-dev libcap-dev libexpat1-dev libfl-dev libpcre3-dev libpcrecpp0 libsigsegv2 libsqlite3-dev libssl-dev libtool m4 po-debconf tcl-dev tcl8.6-dev zlib1g-dev
3) cd trafficserver
4) autoreconf -if
5) ./configure --prefix /opt/ats
6) make all 
7) sudo make install
8) sudo /opt/ats/bin/traffic_manager

Let me know if this helps. I was having the same issue with the latest fast ring build that just installed on my box.

**Comment by misenesi** [comment]
Could you provide an strace of the failed run? Have you noticed any differenced between strace on native Ubuntu and on LxCore?

**Comment by dragon512** [comment]
I have not noticed and differences.. but I have not used it much myself. I just got push of 14931. I going to reboot and rebuild and retry. I will update here once that is finished.

Slightly different issue I will add here is this..

jkenny@DESKTOP-9TV9KQF:~$ ats_62/bin/traffic_manager
[E. Mgmt] log ==> [TrafficManager] using root directory '/home/jkenny/ats_62'
FATAL: ink_sys_control.cc:59: failed assert `setrlimit(MAGIC_CAST(which), &rl) >= 0`
Aborted (core dumped)

This issue is that setrlimit is behaving different on LxCore vs that of Ubuntu. The work around is to run with sudo.

**Comment by dragon512** [comment]
Ok so I have done a lxrun /uninstall /full and then a lxrun /install. I have a clean setup with Build 14931.
Did get some improvement here FYI...

I run these commands:
sudo apt-get install autoconf automake autotools-dev bison debhelper dh-apparmor flex gettext intltool-debian libbison-dev libcap-dev libexpat1-dev libfl-dev libpcre3-dev libpcrecpp0 libsigsegv2 libsqlite3-dev libssl-dev libtool m4 po-debconf tcl-dev tcl8.6-dev zlib1g-dev git
git clone -b 6.2.x https://github.com/apache/trafficserver.git
cd trafficserver
./configure --prefix /opt/ats
make all
make install

The I run:
 jlkenny@DESKTOP-9TV9KQF:~$ **/opt/ats/bin/traffic_manager**
[E. Mgmt] log ==> [TrafficManager] using root directory '/opt/ats'
FATAL: ink_sys_control.cc:59: failed assertion `setrlimit(MAGIC_CAST(which), &rl) >= 0`
Aborted (core dumped)

This is worked around via using sudo:
jlkenny@DESKTOP-9TV9KQF:~/trafficserver$ sudo /opt/ats/bin/traffic_manager
[sudo] password for jlkenny:
[E. Mgmt] log ==> [TrafficManager] using root directory '/opt/ats'
traffic_server: using root directory '/opt/ats'

Now it is running :-) This is good. So now I do a test to get data from the server via running traffic_ctl.
I run:
jlkenny@DESKTOP-9TV9KQF:~$ /opt/ats/bin/traffic_ctl metric get proxy.process.dns.in_flight
proxy.process.dns.in_flight 0

This works. So it **looks like domain sockets are working** :-) I guess you did fix it, It may have been unclear in which drop this fix was in.


So at this point the only issues i seems to have are the issue with rlimit()

The new issue I seem to be having is that ipv6 is not working. In branch 7.0.x ipv6 is set on by default. This shows an issue when running traffic_manager as 

jlkenny@DESKTOP-9TV9KQF:~/trafficserver$ sudo /opt/ats/bin/traffic_manager
[sudo] password for jlkenny:
[E. Mgmt] log ==> [TrafficManager] using root directory '/opt/ats'
FATAL: [bindProxyPort] Unable to bind socket: 8080 : Address already in use

removing from file /opt/ats/etc/trafficserver/records.config the line of 
_CONFIG proxy.config.http.server_ports STRING 8080 8080:ipv6_
to 
_CONFIG proxy.config.http.server_ports STRING 8080_

which turns off ipv6 then it works fine as well. I can open a different issue for this.

**Comment by dragon512** [comment]
Just adding a note that ipv6 seems to work fine given it uses a different port from ipv4. ipv4 and ipv6 should be different port spaces, they should not conflict.

**Comment by sunilmut** [comment]
@dragon512 IPV4 and IPV6 addresses (including ports) can conflict if the IPV6_V6ONLY flag is not used. In the WSL builds that are available to public, there is a issue with the IPV6_V6ONLY (for example see #740). Fortunately, we have been able to fix the IPV6_V6ONLY issue with the help from the Windows networking team, but that fix is not yet publicly available. Meanwhile, to workaround the issue, you can try using different ports for IPV6 and IPV4.

**Comment by dragon512** [comment]
I going to close this issue as it seems to be fixed in the latest build
